Source/DataSource.swift

```diff
@@ -85,77 +93,138 @@ public struct DataSource<Item: ParentType> {
-        guard section < sections.count else { return 0 }
+        guard section < sections.count else {
+            return 0
+        }
         return sections[section].total
     }
-    
+
     public func items(inSection section: Int) -> [Item]? {
-        guard section < sections.count else { return nil }
+        guard section < sections.count else {
+            return nil
+        }
         return sections[section].items
     }
-    
+
     public func item(atRow row: Int, inSection section: Int) -> Item? {
-        guard let items = items(inSection: section) else { return nil }
-        guard row < items.count else { return nil }
+        guard let items = items(inSection: section) else {
+            return nil
+        }
+        guard row < items.count else {
+            return nil
+        }
         return items[row]
     }
-    
+
     public func headerTitle(inSection section: Int) -> String? {
-        guard section < sections.count else { return nil }
+        guard section < sections.count else {
+            return nil
+        }
         return sections[section].headerTitle
     }
-    
+
     public func footerTitle(inSection section: Int) -> String? {
-        guard section < sections.count else { return nil }
+        guard section < sections.count else {
+            return nil
+        }
         return sections[section].footerTitle
     }
-    
+
     public func childItem(atRow row: Int, inSection section: Int, parentIndex: Int, currentPos: Int) -> Item.ChildItem? {
-        guard let items = items(inSection: section) else { return nil }
+        guard let items = items(inSection: section) else {
+            return nil
+        }
         return items[parentIndex].children[row - currentPos - 1]
     }
-    
-    public mutating func expandParent(atIndexPath indexPath: IndexPath, parentIndex: Int) {
-        let section = indexPath.section
-        guard var items = items(inSection: section) else { return }
-        items[parentIndex].state = .expanded
-        sections[section].total += items[parentIndex].children.count
+
+    public mutating func toggleParentCell(toState state: State, inSection section: Int, atIndex index: Int) {
+        guard var parents = items(inSection: section) else {
+            return
+        }
+        parents[index].state = state
+
+        sections[section].total += {
+            var childrenCount = parents[index].children.count
+            if state == .collapsed {
+                childrenCount *= -1
+            }
+            return childrenCount
+        }()
+    }
+
+    public func indexOfFirstExpandedParent() -> IndexPath? {
+        for section in (0..<self.numberOfSections()) {
+            if let parents = items(inSection: section) {
+                for parentIndex in parents.indices {
+                    if parents[parentIndex].state == .expanded {
+                        return IndexPath(item: parentIndex, section: section)
```

Comment:
What do you think if we remove a little of imperative programming using something like this here IMO:
```swift
 return parents
      .firstIndex(where: { $0.state == .expanded })
      .map { IndexPath(item: $0, section: section) }
```

Reply:
I have tried to implement the changes you suggested, I encountered a couple of issues so I will walk you through my thought process.

#### Attempt 1: 
I replaced the inner for loop with the firstIndex. I found that I needed to introduce a variable `indexPath` to allow the first for loop to short circuit if a parent that is expanded is found. I also didn't like the mix of imperial and functional programming. That lead me to attempt2.
```swift
public func indexOfFirstExpandedParent() -> IndexPath? {
   var indexPath: IndexPath? = nil
   for section in (0..<self.numberOfSections()) {
      if let parents = items(inSection: section) {
          return parents.firstIndex(where: { $0.state == .expanded })
                        .map { indexPath = IndexPath(item: $0, section: section) }
      }
      
      if indexPath == nil {
        return indexPath
      }
  }
  return nil
}
```

#### Attempt 2:
I decided to go with an all functional approach. I encountered the following issues:
- In the below code the reference to the section in which the parent belongs gets lost in the flat map and thus the indexpath produced is incorrect.


```swift
public func indexOfFirstExpandedParent() -> IndexPath? {
    (0..<self.numberOfSections())
            .map { section in items(inSection: section) }
            .flatMap{ $0 ?? []}
            .firstIndex(where: { $0.state == .expanded })
            .map { IndexPath(item: $0, section: // Lost in flatmap) }
}
```

I then decided to try wrap the parents and section in a tuple. Please see the below signatures:
```
map:     [Int] -> ([Item]?, Int)
flatMap: [([Item]?, Int)] -> [(Item, Int)] // I had trouble getting this correct conversion working correctly in a clean way
first:   [(Item, Int)] -> (Item, Int)      // (where state == expanded )
map:     (Item, Int) -> IndexPath?
```
I was not able to get the flatmap to return what I wanted, I thus have left the implementation of Attempt 1. 
Please feel free to point out if I missed something obvious, or if you have a better solution.

Reply:
Thank you for the response. I agree that we don't need the index path variable, I have removed it from the below code.  Does this look more correct?
```swift
public func indexOfFirstExpandedParent() -> IndexPath? {
    // Loop through each section if there are multiple
    for section in (0..<self.numberOfSections()) {
        if let parents = items(inSection: section) {
            // Find parent that is expanded
            let indexPath = parents.firstIndex(where: { $0.state == .expanded })
                                   .map { IndexPath(item: $0, section: section) }

            // Return the index path if a parent is found to be expanded
            if indexPath != nil {
                return indexPath
            }
        }
    }
    return nil
}
```

Please see the changes to the PR, I have updated the code to the above.

Reply:
Yep, that looks very good!